core/iwasm/libraries/lib-rats/lib_rats_wrapper.c

```diff
@@ -8,38 +8,79 @@
-librats_collect_wrapper(wasm_exec_env_t exec_env, const uint8_t *hash)
+extern char wasm_module_hash[SHA256_DIGEST_LENGTH];
+
+static int
+librats_collect_wrapper(wasm_exec_env_t exec_env, rats_sgx_evidence_t *evidence,
+                        uint32_t evidence_size, const char *buffer,
+                        uint32_t buffer_size)
 {
-    char *json = NULL;
-    char *str_ret;
-    uint32 len;
-    uint32 str_ret_offset = 0;
-    wasm_module_inst_t module_inst = get_module_inst(exec_env);
-    int code = librats_collect_evidence_to_json(hash, &json);
-    if (code != 0) {
-        return str_ret_offset;
-    }
-    if (json) {
-        len = (uint32)strlen(json) + 1;
+    attestation_evidence_t att_ev;
 
-        str_ret_offset = module_malloc(len, (void **)&str_ret);
-        if (str_ret_offset) {
-            bh_memcpy_s(str_ret, len, json, len);
-        }
+    char final_hash[SHA256_DIGEST_LENGTH] = { '\0' };
+    SHA256_CTX sha256;
+    SHA256_Init(&sha256);
+    SHA256_Update(&sha256, wasm_module_hash, SHA256_DIGEST_LENGTH);
+    if (buffer != NULL)
+        SHA256_Update(&sha256, buffer, strlen(buffer));
+    SHA256_Final((unsigned char *)final_hash, &sha256);
+
+    rats_attester_err_t ret_code =
+        librats_collect_evidence(&att_ev, (const uint8_t *)final_hash);
+    if (ret_code != RATS_ATTESTER_ERR_NONE) {
+        return (int)ret_code;
     }
-    return str_ret_offset;
+
+    sgx_quote3_t *quote_ptr = (sgx_quote3_t *)att_ev.ecdsa.quote;
+    bh_memcpy_s(evidence->quote, att_ev.ecdsa.quote_len, att_ev.ecdsa.quote,
+                att_ev.ecdsa.quote_len);
+    evidence->quote_size = att_ev.ecdsa.quote_len;
+    bh_memcpy_s(evidence->user_data, SGX_REPORT_DATA_SIZE,
+                quote_ptr->report_body.report_data.d, SGX_REPORT_DATA_SIZE);
+    bh_memcpy_s(evidence->mr_enclave, sizeof(sgx_measurement_t),
+                quote_ptr->report_body.mr_enclave.m, sizeof(sgx_measurement_t));
+    bh_memcpy_s(evidence->mr_signer, sizeof(sgx_measurement_t),
+                quote_ptr->report_body.mr_signer.m, sizeof(sgx_measurement_t));
+    evidence->product_id = quote_ptr->report_body.isv_prod_id;
+    evidence->security_version = quote_ptr->report_body.isv_svn;
+    evidence->att_flags = quote_ptr->report_body.attributes.flags;
+    evidence->att_xfrm = quote_ptr->report_body.attributes.flags;
+
+    return 0;
 }
 
 static int
-librats_verify_wrapper(wasm_exec_env_t exec_env, const char *evidence_json,
-                       const uint8_t *hash)
+librats_verify_wrapper(wasm_exec_env_t exec_env, rats_sgx_evidence_t *evidence,
+                       uint32_t evidence_size, const char *buffer,
+                       uint32_t buffer_size)
 {
-    return librats_verify_evidence_from_json(evidence_json, hash);
+    attestation_evidence_t att_ev;
+
+    char final_hash[SHA256_DIGEST_LENGTH] = { '\0' };
+    SHA256_CTX sha256;
+    SHA256_Init(&sha256);
+    SHA256_Update(&sha256, wasm_module_hash, SHA256_DIGEST_LENGTH);
```

Comment:
In the case of verification, shouldn't we allow to receive `wasm_module_hash` by a parameter of the _verify_ function? Otherwise, a given Wasm module cannot verify the evidence of a different Wasm module. It will only be able to verify the same module (the same module must collect and verify). Correct me if I'm wrong.

Reply:
Thanks for your review and pointing out this mistake. And I realize it's incorrect to generate structured evidence in `librats_collect` because we can't verify the integrity of it and it's not consistent with the structure in `librats`.
Following is my modification now:

- `librats_collect` generate the json of evidence with the structure `attestation_evidence_t` in `librats`, it will take `wasm_module_hash` into `user_data` of the quote.
- `librats_verify` verify the json of evidence with the structure `attestation_evidence_t` in `librats`, it just accept the overall hash like the value in `user_data` of the quote as an input parameter.
- `librats_parse_evidence ` parse the json of evidence with the structure `attestation_evidence_t` in `librats` into `rats_sgx_evidence`, we can use it for further verification. It only support parse sgx evidence now.

Thanks again for your review. Could you help review it again? Thanks.

Reply:
Great idea to have introduced `librats_parse_evidence`. This makes sense when working with the evidence, thanks!